source/gameengine/Device/DEV_JoystickVibration.cpp

```diff
@@ -0,0 +1,190 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ ** Contributor(s): youle, lordloki
+ *
+ * ***** END GPL LICENSE BLOCK *****
+ */
+
+/** \file gameengine/Device/DEV_JoystickVibration.cpp
+ *  \ingroup device
+ */
+
+#include "DEV_Joystick.h"
+#include "DEV_JoystickPrivate.h"
+#include "DEV_JoystickDefines.h"
+
+#include <memory> // We have to include that on Windows to make memset available
+
+bool DEV_Joystick::RumblePlay(float strength_left, float strength_right, unsigned int duration)
+{
+#ifdef WITH_SDL
+	unsigned int effects;
+	bool run_by_effect = false;
+	bool effects_issue = false;
+
+	if (m_private->m_haptic == NULL) {
+		return false;
+	}
+
+	// Managing vibration logic
+	if (m_private->m_hapticeffect_status == JOYHAPTIC_STOPPED) {
+		memset(&m_private->m_hapticeffect, 0, sizeof(SDL_HapticEffect)); // 0 is safe default
+	}
+	else if (m_private->m_hapticeffect_status == JOYHAPTIC_PLAYING_EFFECT) {
+		m_private->m_hapticeffect_status = JOYHAPTIC_UPDATING_EFFECT;
+	}
+	else if (m_private->m_hapticeffect_status == JOYHAPTIC_PLAYING_RUMBLE) {
+		m_private->m_hapticeffect_status = JOYHAPTIC_UPDATING_RUMBLE;
+	}
+
+	// Checking supported effects
+	effects = SDL_HapticQuery(m_private->m_haptic);
+
+	if ((effects & SDL_HAPTIC_LEFTRIGHT) && m_private->m_hapticeffect_status != JOYHAPTIC_UPDATING_RUMBLE) {
+		if (m_private->m_hapticeffect_status != JOYHAPTIC_UPDATING_EFFECT) {
+			m_private->m_hapticeffect.type = SDL_HAPTIC_LEFTRIGHT;
+		}
+
+		m_private->m_hapticeffect.leftright.length = duration;
+		m_private->m_hapticeffect.leftright.large_magnitude = (unsigned int) (strength_left * 0x7FFF);
+		m_private->m_hapticeffect.leftright.small_magnitude = (unsigned int) (strength_right * 0x7FFF);
+		run_by_effect = true;
+
+	}
+	// Some Game Controllers only supports large/small magnitude motors using a custom effect
+	else if ((effects & SDL_HAPTIC_CUSTOM) && m_private->m_hapticeffect_status != JOYHAPTIC_UPDATING_RUMBLE) {
+
+		Uint16 data[2]; // data = channels * samples
+		data[0] = (Uint16) (strength_left * 0x7FFF);
+		data[1] = (Uint16) (strength_right * 0x7FFF);
+
+		if (m_private->m_hapticeffect_status != JOYHAPTIC_UPDATING_EFFECT) {
+			m_private->m_hapticeffect.type = SDL_HAPTIC_CUSTOM;
+		}
+		m_private->m_hapticeffect.custom.length = duration;
+		m_private->m_hapticeffect.custom.channels = 2;
+		m_private->m_hapticeffect.custom.period = 1;
+		m_private->m_hapticeffect.custom.samples = 1;
+		m_private->m_hapticeffect.custom.data = data;
+
+		run_by_effect = true;
+	}
+
+	if (run_by_effect) {
+		bool new_effect = true;
+
+		if (m_private->m_hapticeffect_status == JOYHAPTIC_UPDATING_EFFECT) {
+			if (SDL_HapticUpdateEffect(m_private->m_haptic, m_private->m_hapticeffect_id, &m_private->m_hapticeffect) == 0) {
+				m_private->m_hapticeffect_status = JOYHAPTIC_PLAYING_EFFECT;
+				new_effect = false;
+			}
+			else {
+				SDL_HapticDestroyEffect(m_private->m_haptic, m_private->m_hapticeffect_id);
```

Comment:
This statement is for an error ? I yes, you can add an error message with `CM_Error`